src/assign.c

```diff
@@ -811,322 +797,201 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source,
-          }
-        }
-      } else {
-        if (TYPEOF(source)!=STRSXP) error("Internal error. Not coerced earlier.");
-        if (slen==1) {
-          const SEXP val = STRING_ELT(source, 0);
-          for (int i=0; i<len; i++) SET_STRING_ELT(target, start+i, val);
-        } else {
-          const SEXP *val = STRING_PTR(source);
-          for (int i=0; i<len; i++) SET_STRING_ELT(target, start+i, val[i]);
-        }
-      }
-      break;
-    case VECSXP :
-      if (TYPEOF(source)==VECSXP && len==slen) {
-        for (int i=0; i<len; i++) SET_VECTOR_ELT(target, start+i, VECTOR_ELT(source, i));
-      } else {
-        const SEXP val = TYPEOF(source)==VECSXP ? VECTOR_ELT(source, 0) : source;
-        for (int i=0; i<len; i++) SET_VECTOR_ELT(target, start+i, val);
-      }
-      break;
-    default :
-      error("Unsupported type in assign.c:memrecycle '%s' (no where)", type2char(TYPEOF(target)));  // # nocov
+
+#undef BODY
+#define BODY(STYPE, RFUN, CTYPE, CAST, ASSIGN) \
+  { const STYPE *sd = (const STYPE *)RFUN(source);   \
+  if (length(where)) {                         \
+    const int *wd = INTEGER(where)+start;      \
+    if (slen==1) {                             \
+      const STYPE val = sd[0];                 \
+      const CTYPE cval = CAST;                 \
+      for (int wi=0; wi<len; ++wi) {           \
+        const int w = wd[wi];                  \
+        if (w<1) continue; /*0 or NA*/         \
+        const int i = w-1;                     \
+        ASSIGN;                                \
+      }                                        \
+    } else {                                   \
+      for (int wi=0; wi<len; ++wi) {           \
+        const int w = wd[wi];                  \
+        if (w<1) continue;                     \
+        const STYPE val = sd[wi];              \
+        const CTYPE cval = CAST;               \
+        const int i = w-1;                     \
+        ASSIGN;                                \
+      }                                        \
+    }                                          \
+  } else {                                     \
+    if (slen==1) {                             \
+      const STYPE val = sd[0];                 \
+      const CTYPE cval = CAST;                 \
+      for (int i=0; i<len; ++i)                \
+        ASSIGN;                                \
+    } else {                                   \
+      for (int i=0; i<len; i++) {              \
+        const STYPE val = sd[i];               \
+        const CTYPE cval = CAST;               \
+        ASSIGN;                                \
+      }                                        \
+    }                                          \
+  }}
+
+#define COERCE_ERROR(targetType) error("type '%s' cannot be coerced to '%s'", type2char(TYPEOF(source)), targetType); // 'targetType' for integer64 vs double
```

Comment:
shouldn't be `source` also an argument to `COERCE_ERROR`, it iss being used inside

Reply:
OK, I understand it is taken from parent scope in `switch` and I assume it never changes

Reply:
yep this is a little inelegant.  it was because the type of REALSXP doesn't distinguish between "double" and "integer64", which is what I was trying to say by the comment alongside.  We probably do have a slightly higher level function superType() or something which does this.  If not we should create it and replace all type2char(TYPEOF(...)) with superType() instead.